include/Homa/Core/Transport.h

```diff
@@ -0,0 +1,162 @@
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+/**
+ * @file Homa/Core/Transport.h
+ *
+ * Contains the low-level Homa Transport API.  Advanced users of the Homa
+ * Transport library should include this header.
+ */
+
+#pragma once
+
+#include <Homa/Homa.h>
+
+namespace Homa::Core {
+
+/**
+ * Minimal set of low-level API that can be used to create Homa-based transports
+ * for different runtime environments (e.g. polling, kernel threading,
+ * green threads, etc).
+ *
+ * The execution of a transport is driven through repeated calls to methods
+ * like checkTimeouts(), processPacket(), trySend(), and trySendGrants(); the
+ * transport will not make any progress otherwise. Advanced users can compose
+ * these methods in a way that suits them best.
+ *
+ * This class is thread-safe.
+ */
+class Transport : public TransportBase {
+  public:
+    /**
+     * Collection of user-defined transport callbacks.
+     */
+    class Callbacks {
+      public:
+        /**
+         * Destructor.
+         */
+        virtual ~Callbacks() = default;
+
+        /**
+         * Invoked when an incoming message arrives and needs to dispatched to
+         * its destination in the user application for processing.
+         *
+         * Here are a few example use cases of this callback:
+         * <ul>
+         * <li> Interaction with the user's thread scheduler: e.g., an
+         * application may want to block on receive until a message is
+         * delivered, so this method can be used to wake up blocking threads.
+         * <li> High-performance message dispatch: e.g., an application may
+         * choose to implement the message receive queue with a concurrent MPMC
+         * queue as opposed to a linked-list protected by a mutex; <li>
+         * Lightweight synchronization: e.g., the socket table that maps from
+         * port numbers to sockets is a read-mostly data structure, so lookup
+         * operations can benefit from synchronization schemes such as RCU.
+         * </ul>
+         *
+         * @param port
+         *      Destination port number of the message.
+         * @param message
+         *      Incoming message to dispatch.
+         * @return
+         *      True if the message is delivered successfully; false, otherwise.
+         */
+        virtual bool deliver(uint16_t port, InMessage* message) = 0;
+
+        /**
+         * Invoked when some packets just became ready to be sent (and there was
+         * none before).
+         *
+         * This callback allows the transport library to notify the users that
+         * trySend() should be invoked again as soon as possible. For example,
+         * the callback can be used to implement wakeup signals for the thread
+         * that is responsible for calling trySend(), if this thread decides to
+         * sleep when there is no packets to send.
+         */
+        virtual void notifySendReady() {}
```

Comment:
There was one but John convinced me that using a separate grantor thread is probably not a good idea. I have added some documentation to `trySendGrants` to give an example on how it could be used.